db/migrate/20241030122707_create_note_tags.rb

```diff
@@ -0,0 +1,12 @@
+class CreateNoteTags < ActiveRecord::Migration[7.2]
+  def change
+    # Create a table and primary key
+    create_table :note_tags, :primary_key => [:note_id, :k] do |t|
+      t.column "note_id", :bigint, :null => false
+      t.column "k",  :string, :default => "", :null => false
+      t.column "v",  :string, :default => "", :null => false
+
+      t.foreign_key :notes, :column => :note_id, :name => "note_tags_id_fkey"
+    end
+  end
+end
```

Comment:
This migration works well for ensuring uniqueness with a composite primary key. However, if querying by tags (e.g., `{k: "created_by", v: "openstreetmap"}`) is expected, a surrogate primary key (`id`) with additional indexes might be more efficient. For example:

```ruby
create_table :note_tags do |t|
  t.bigint :note_id, null: false
  t.string :k, null: false, default: ""
  t.string :v, null: false, default: ""

  t.timestamps
end

add_index :note_tags, [:note_id, :k], unique: true       # Enforce uniqueness
add_index :note_tags, [:k, :v]                          # Optimize tag-based queries
```

This approach aligns with Rails conventions, simplifies queries, and allows adding indexes for tag lookups (`k`/`v`).

Reply:
How is a surrogate key going to help with queries?

Reply:
I misunderstood the use case, you are right it doesn't. Adding the index on k and v would do.. I will edit the top comment for more clarity